Bottom: c7b97f58a0946ddd6f2f158b274b7b4b515b5387
Top:    339971121c9e29c477cbb2f30cf6dbebc3be20ff
Author: Jason Wang <jasowang@redhat.com>
Date:   2018-02-28 14:57:39 +0800

Refresh of virtio-net-re-enable

---

diff --git a/drivers/net/virtio_net.c b/drivers/net/virtio_net.c
index a94319a..0f12cf2 100644
--- a/drivers/net/virtio_net.c
+++ b/drivers/net/virtio_net.c
@@ -710,6 +710,7 @@ static struct sk_buff *receive_mergeable(struct net_device *dev,
 	unsigned int truesize;
 	unsigned int headroom = mergeable_ctx_to_headroom(ctx);
 	bool sent, skb_xdp = false;
+	int err;
 
 	head_skb = NULL;
 
@@ -772,6 +773,13 @@ static struct sk_buff *receive_mergeable(struct net_device *dev,
 			*xdp_xmit = true;
 			rcu_read_unlock();
 			goto xdp_xmit;
+		case XDP_REDIRECT:
+			err = xdp_do_redirect(dev, &xdp, xdp_prog);
+			if (err)
+				goto err_xdp;
+			*xdp_xmit = true;
+			rcu_read_unlock();
+			goto xdp_xmit;
 		default:
 			bpf_warn_invalid_xdp_action(act);
 		case XDP_ABORTED:
